SYCL/InorderQueue/in_order_usm_explicit.cpp

```diff
@@ -0,0 +1,64 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// SYCL in ordered queues explicit USM test
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <CL/sycl.hpp>
+#include <iostream>
+
+using namespace cl::sycl;
+
+int main() {
+
+  {
+    const int dataSize = 32;
+    const size_t numBytes = static_cast<size_t>(dataSize) * sizeof(int);
+
+    int dataA[dataSize] = {0};
+    int dataB[dataSize] = {0};
+
+    queue Queue{property::queue::in_order()};
+
+    auto devicePtrA = malloc_device<int>(numBytes, Queue);
+    Queue.memcpy(devicePtrA, &dataA, numBytes);
+
+    Queue.submit([&](handler &cgh) {
+      auto myRange = range<1>(dataSize);
+      auto myKernel = ([=](id<1> idx) { devicePtrA[idx] = idx[0]; });
+
+      cgh.parallel_for<class ordered_writer>(myRange, myKernel);
+    });
+
+    auto devicePtrB = malloc_device<int>(numBytes, Queue);
+    Queue.memcpy(devicePtrB, &dataB, numBytes);
+
+    Queue.submit([&](handler &cgh) {
+      auto myRange = range<1>(dataSize);
+      auto myKernel = ([=](id<1> idx) { devicePtrB[idx] = devicePtrA[idx]; });
+
+      cgh.parallel_for<class ordered_reader>(myRange, myKernel);
+    });
+
+    Queue.memcpy(&dataB, devicePtrB, numBytes);
+
+    Queue.wait();
+
+    auto ctxt = Queue.get_context();
+    free(devicePtrA, ctxt);
+    free(devicePtrB, ctxt);
+
+    for (int i = 0; i != dataSize; ++i) {
+      if (dataB[i] != i) {
+        std::cout << "Result mismatches " << dataB[i] << " vs expected " << i
+                  << " for index " << i << std::endl;
+      }
+    }
```

Comment:
Is it tested that for `queue` with no `in_order` property the result can mismatch eventually?

Reply:
Good point.  I don't see that this question has been mentioned before in the commit history of in_order queue tests from before the tests were migrated from llvm to the llvm-test-suite.  I tried removing the in_order property from tests in the InorderQueue folder and running them on different devices.  In all cases I see that the tests pass, which seems surprising.  At least for CUDA I don't think that the use of the in_order property is currently linked to the backend so no difference in behaviour is expected when using this property.  For CUDA I think that only in_order queues(streams) are currently implemented so it is not possible to compare the behaviour between in_order to default queues for CUDA.  I don't know whether this is also true for other cases.

Perhaps all in_order queue tests should be adapted so that they are expected to fail in all cases that the in_order property is not used.

Reply:
> 
> 
> Is it tested that for `queue` with no `in_order` property the result can mismatch eventually?

As @JackAKirk mentions, the CUDA backend does not currently support out-of-order execution, so in-order is assumed even if told otherwise in the properties.

However, even if it wasn't the case, testing that out-of-order queues actually do work out of order is tough for a handful of reasons where most are hardware specific, e.g. even something as simple as having a faster GPU may defer any value mismatches or - depending on how the test was written - never get to a stage where the values wouldn't mismatch at all. My fear would be that adding such a test would either have to have a very long timeout, making it annoying when it actually fails, or be short but with a chance that it would sporadically fail. Neither seems like a good option.

Reply:
Is it okay, then, to verify that correct properties are passed to backend using output of the app running with `SYCL_PI_TRACE=-1`?

Reply:
Interesting idea. Are there other tests that do something similar? I believe the properties are passed to the corresponding PI functions in a bitfield, so it would need to parse the correct parameter value and check the corresponding bit.

Reply:
One can `grep` the tests for `SYCL_PI_TRACE` and `FileCheck` to find the tests that do smth similar.

Reply:
Does FileCheck allow for parsing? If we want to check that the property is set, we will have to check the corresponding bit in the parameter, otherwise if we compare it to a static value it would start to fail if other properties are implicitly passed in the future, even though the out-of-order property bit stays set.

Reply:
The manual says, regexps are allowed to some extent.
Anyway, a static value is still an option i.e. in case of implicit properties been added, one should fix the test. The test here, should have a properly understandable comment.

Reply:
A test for the correct queue property in piQueueCreate has been made: https://github.com/intel/llvm-test-suite/pull/272